src/Domain/CoordinateMaps/FocallyLiftedEndcap.cpp

```diff
@@ -0,0 +1,331 @@
+
+  return jacobian_matrix;
+}
+
+template <typename T>
+tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame> Endcap::inv_jacobian(
+    const std::array<T, 3>& source_coords) const noexcept {
+  using ReturnType = tt::remove_cvref_wrap_t<T>;
+  const ReturnType& xbar = source_coords[0];
+  const ReturnType& ybar = source_coords[1];
+  const ReturnType rhobar = sqrt(square(xbar) + square(ybar));
+  // Let q = sin(rhobar theta)/rhobar
+  const ReturnType q = Endcap_detail::sin_ax_over_x(rhobar, theta_max_);
+  const ReturnType dlogrhobar_q =
+      Endcap_detail::dlogx_sin_ax_over_x(rhobar, theta_max_);
+  const ReturnType one_over_r_q = 1.0 / (q * radius_);
+  const ReturnType jacobian_factor =
+      one_over_r_q * dlogrhobar_q / (q + square(rhobar) * dlogrhobar_q);
+
+  auto inv_jacobian_matrix =
+      make_with_value<tnsr::Ij<tt::remove_cvref_wrap_t<T>, 3, Frame::NoFrame>>(
+          dereference_wrapper(source_coords[0]), 0.0);
+
+  // dxbar/dx
+  get<0, 0>(inv_jacobian_matrix) =
+      one_over_r_q - square(xbar) * jacobian_factor;
+  // dxbar/dy
+  get<0, 1>(inv_jacobian_matrix) = -xbar * ybar * jacobian_factor;
+  // dybar/dx
+  get<1, 0>(inv_jacobian_matrix) = get<0, 1>(inv_jacobian_matrix);
+  // dybar/dy
+  get<1, 1>(inv_jacobian_matrix) =
+      one_over_r_q - square(ybar) * jacobian_factor;
+
+  return inv_jacobian_matrix;
+}
+
+boost::optional<std::array<double, 3>> Endcap::inverse(
+    const std::array<double, 3>& target_coords,
+    const double sigma_in) const noexcept {
+  const double x = target_coords[0] - center_[0];
+  const double y = target_coords[1] - center_[1];
+  const double z = target_coords[2] - center_[2];
+
+  // Are we in the range of the map?
+  // The equal_within_roundoff below has an implicit scale of 1,
+  // so the inverse may fail if radius_ is very small on purpose,
+  // e.g. if we really want a tiny tiny domain for some reason.
+  const double r = sqrt(square(x) + square(y) + square(z));
+  if (not equal_within_roundoff(r, radius_)) {
+    return boost::none;
+  }
+
+  // Compute zbar and check if we are in the range of the map.
+  const double zbar = 2.0 * sigma_in - 1.0;
+  if (abs(zbar) > 1.0 and not equal_within_roundoff(abs(zbar), 1.0)) {
+    return boost::none;
+  }
+
+  const double rho = sqrt(square(x) + square(y));
+  if (UNLIKELY(rho == 0.0)) {
+    // If x and y are zero, so are xbar and ybar,
+    // so we are done.
+    return std::array<double, 3>{{0.0, 0.0, zbar}};
+  }
+
+  // Note: theta_max_ cannot be zero for a nonsingular map.
+  const double rhobar = atan2(rho, z) / theta_max_;
+
+  // Check if we are outside the range of the map.
+  if (rhobar > 1.0 and not equal_within_roundoff(rhobar, 1.0)) {
+    return boost::none;
+  }
+
+  const double xbar = x * rhobar / rho;
+  const double ybar = y * rhobar / rho;
+  return std::array<double, 3>{{xbar, ybar, zbar}};
+}
+
+template <typename T>
+tt::remove_cvref_wrap_t<T> Endcap::sigma(
```

Comment:
I would recommend switch all these internal helpers to returning by `not_null`. I'm worried that all the memory allocations will mean someone else will need to do that anyway in the future, plus you might be able to eliminate some allocations immediately.